### ISSUE TITLE
nullable<alias> does generate nullable type for decimal

### ISSUE THREAD

**Issue by DanielFabian** [Microsoft/bond#590]
I have a schema, where I define decimal as in the example and it correctly generates the `decimal` in C#. But if the property is a `nullable<decimal>` it does not work for some reason. The same works just fine for `DateTime`. This schema:

    namespace Repro
    
    using decimal = blob;
    
    struct ReproStruct
    {    
        0: decimal MyProp;
        1: nullable<decimal> MyBuggyProp;
    }

generates this faulty C#:

    namespace Repro
    {
        using System.Collections.Generic;

        [global::Bond.Schema]
        [System.CodeDom.Compiler.GeneratedCode("gbc", "0.10.0.0")]
        public partial class ReproStruct
        {
            [global::Bond.Id(0), global::Bond.Type(typeof(global::Bond.Tag.blob))]
            public decimal MyProp { get; set; }

            [global::Bond.Id(1), global::Bond.Type(typeof(global::Bond.Tag.nullable<global::Bond.Tag.blob>))]
            public decimal MyBuggyProp { get; set; }

            public ReproStruct()
                : this("Repro.ReproStruct", "ReproStruct")
            {}

            protected ReproStruct(string fullName, string name)
            {
                MyProp = new decimal();
            }
        }
    } // Repro

**Comment by chwarr** [comment]
Yup, I see the problem. Thanks for the report.

When [determining the C# type to emit for a nullable field](https://github.com/Microsoft/bond/blob/fcd492dfd7870774e604130ab9f2d8dc1070da8b/compiler/src/Language/Bond/Codegen/TypeMapping.hs#L422), either `type` (for a C# reference type) or `type?` (for a C# value type), gbc uses the [`isScalar`](https://github.com/Microsoft/bond/blob/939c298ff6d619dbc0bc9f5ca25c2088de86d83b/compiler/src/Language/Bond/Syntax/Util.hs#L49-L65) function, and appends the `?` for scalar types. (For aliases, `isScalar` returns the scalarness of the underlying Bond type.) `isScalar` operates in the Bond type system, independent of the target language. This works out when `isScalar` and "Is a C# value type?" are the same, which is the case when no type mapping is going on.

The alias `using decimal=blob;` with `--using=decimal=decimal` caused `isScalar` and "Is a C# value type?" to get out of sync. A Bond blob is non-scalar, so no `?` is appended, but a `decimal`/`System.Decimal` is a value type, so it needs to be `decimal?`/`Nullable<decimal>`/... to permit nulls.

As a workaround, you can use something like this:

    namespace Repro;
    using decimal=blob;
    using ndecimal=blob;

    struct HasNullableDecimal
    {
        0: decimal nonNullableDecimal;
        1: nullable<ndecimal> nullableDecimal;
    }

and run gbc with `--using="decimal=decimal" --using="ndecimal=decimal?"`.

The reverse problem can happen as well. Consider the following .bond file:

    namespace Repro;
    using DateTime=int64;

    struct HasNullableDateTime
    {
        0: nullable<DateTime> nullableDateTime;
    }

When gbc is run with `--using="DateTime=System.DateTime"`, the following invalid C# is produced. Notice that `System.DateTime?` was generated, but that is not valid, as System.DateTime is a reference type and already nullable.

```csharp
//   Tool : Bond Compiler 0.10.0.0

namespace Repro
{
    using System.Collections.Generic;

    [global::Bond.Schema]
    [System.CodeDom.Compiler.GeneratedCode("gbc", "0.10.0.0")]
    public partial class HasNullableDateTime
    {
        [global::Bond.Id(0), global::Bond.Type(typeof(global::Bond.Tag.nullable<long>))]
        public System.DateTime? nullableDateTime { get; set; }

        public HasNullableDateTime()
            : this("Repro.HasNullableDateTime", "HasNullableDateTime")
        {}

        protected HasNullableDateTime(string fullName, string name)
        {
            
        }
    }
} // Repro
```

To make this work robustly, I see a couple of options, none of which are particularly satisfying...

1. Somehow have gbc know whether the final C# type is a reference type or a value type and pick `Nullable<T>`/`T` based on that knowledge.
    * The referenceness/valueness of small set of well-known C# types could be included in gbc, but there would need to be some way for users of Bond to provide information about other types.
        * This would probably be a command line argument.
1. Instead of projecting `nullable<T>` fields into C# as either `Nullable<T>` or a bare `T`, introduce a type like `BondNullable<T>` that can hold both reference and value types.
    * Either this breaks all existing C# code that manipulates nullable fields, as the type will change, or gbc and the C# library have two different code paths for `Nullable<T>`/`T` and `BondNullable<T>`.
        * The `Nullable<T>`/`T` code path could potentially be removed many versions later.
1. Explicitly do not support mapping to value types outside of byte, int, unit, double, float, enum, &c.
    * This removes functionality that already exists and would break existing code.

Architecturally, number 2 is the cleanest, but the breaking change is too big. Thus, number 1 is the solution I recommend. Anyone else have other suggestions, pros/cons?

**Comment by DanielFabian** [comment]
I like the idea of command line arguments, as in c# the value type vs. reference type is purely a programmer decision. And it feels quite non-intrusive, since the `--using` has to be done manually today already anyway. Adding another argument like `--usingAsValueType` should be a very small change for people using it.

By the way, `System.DateTime` actually __is__ a nullable type, so the generated code is correct.

**Comment by chadwalters** [comment]
I agree that 1 is the best option of the 3. Given that the problem can happen in both directions, we need to be able to explicitly mark both value and reference .

**Comment by chwarr** [comment]
Thanks for pointing out the mistake with DateTime. I've [updated](https://github.com/Microsoft/bond/issues/590#issuecomment-327659711) the example to use System.Tuple instead.
